backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/ColumnarCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.backendsapi.velox
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.backendsapi.BackendsApiManager
+import io.glutenproject.columnarbatch.ColumnarBatches
+import io.glutenproject.execution.RowToVeloxColumnarExec
+import io.glutenproject.memory.arrowalloc.ArrowBufferAllocators
+import io.glutenproject.memory.nmm.NativeMemoryManagers
+import io.glutenproject.utils.{FieldRecordBatch, VectorLoaderWithField, VectorUnloaderWithField}
+import io.glutenproject.utils.ArrowRecordBatchType.{FieldBuffers, FieldNumRows, FiledNullCounts}
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.{InternalRow, SQLConfHelper}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.columnar.{CachedBatch, CachedBatchSerializer}
+import org.apache.spark.sql.execution.VeloxColumnarToRowExec
+import org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{StructField, StructType}
+import org.apache.spark.sql.utils.SparkArrowUtil
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.TaskResources
+
+import org.apache.arrow.vector.VectorSchemaRoot
+
+case class CachedColumnarBatch(
+    override val numRows: Int,
+    override val sizeInBytes: Long,
+    fieldBuffersArr: Array[FieldBuffers],
+    fieldNullCountsArr: Array[FiledNullCounts],
+    fieldNumRowsArr: Array[FieldNumRows])
+  extends CachedBatch {}
```

Comment:
Thanks for working on this!

A question along with the PR description

> Spark does not support cache a offheap object, and it converts UnsafeRow to
Array[Byte] then cache it. So this serializer also use that way, by converting velox ColumnarBatch to
Array[Byte] then cache it.

Since Spark's developer API `CachedBatch` doesn't enforce a format for the underlying data, was there any special reason Array[Byte] is still used in the PR?

Reply:
yeah, there is no constraint, but it seems Array[Byte] is a good option. Say, arrow based buf is offheap and we need a onheap buf.